### DISCUSSION TITLE
example many-to-many insert

### DISCUSSION THREAD

**Discussion by Deliana90** [geldata/gel#4861] [Q&A]
How to do [create Post and CategoriesOnPosts](https://www.prisma.io/docs/concepts/components/prisma-schema/relations/many-to-many-relations#querying-an-explicit-many-to-many) in edgedb? can you give example using typescript please, thanks.

**Comment by raddevon** [comment]
I'll start with the schema:
```
module default {
  type Post {
    required property title -> str;
    multi link categories -> Category;
  }
  type Category {
    required property name -> str;
  }
}
```

A couple of notes on the schema as compared to that in the Prisma example:
- You don't need to create the IDs. EdgeDB handles that for you.
- You don't need the intermediary table which is `CategoriesOnPosts` in this example.

With that schema, you can insert a post with new categories using this code in TypeScript:

```ts
const client = createClient();
const insertPostQuery = e.insert(e.Post, {
  title: "How to be Bob",
  categories: e.set(
    e.insert(e.Category, { name: "Category 1" }),
    e.insert(e.Category, { name: "Category 2" })
  ),
});
await insertPostQuery.run(client);
```

That's what you'd do if you needed to insert new categories. If `Category 1` and `Category 2` already exist, you could do this instead:
```ts
const client = createClient();
const insertPostQuery = e.insert(e.Post, {
  title: "How to be Bob Part 2",
  categories: e.set(
    e.select(e.Category, () => ({ filter_single: { name: "Category 1" } })),
    e.select(e.Category, () => ({ filter_single: { name: "Category 2" } }))
  ),
});
await insertPostQuery.run(client);
```

If you have existing categories and know the IDs of those categories rather than the names, you could do this instead:
```ts
const client = createClient();
const insertPostQuery = e.insert(e.Post, {
  title: "How to be Bob Part 2",
  categories: e.set(
    e.select(e.Category, () => ({ filter_single: { id: "<category-1-id>" } })),
    e.select(e.Category, () => ({ filter_single: { id: "<category-2-id>" } }))
  ),
});
await insertPostQuery.run(client);
```
replacing `<category-1-id>` and `<category-2-id>` with those actual IDs.

If you don't know whether the categories exist, you could do this:
```ts
const client = createClient();
const insertPostQuery = e.insert(e.Post, {
  title: "How to be Bob Part 2",
  categories: e.set(
    e.insert(e.Category, { name: "Category 1" }).unlessConflict((category) => ({on: category.name, else: category})),
    e.insert(e.Category, { name: "Category 2" }).unlessConflict((category) => ({on: category.name, else: category})),
  ),
});
await insertPostQuery.run(client);
```

Note that, as @Syzuna noted in their reply, `unlessConflict` would required you change the schema to make the category name exclusive. That would look like this:

```
module default {
  type Post {
    required property title -> str;
    multi link categories -> Category;
  }
  type Category {
    required property name -> str {
      constraint exclusive;
    }
  }
}
```

I hope this helps!

**Reply by Syzuna** [reply]
Wouldn't you need to adjust your schema for the last example? More specifically add an unique constraint on the name for unless conflict to work?

**Reply by raddevon** [reply]
@Syzuna Yes, you're right. Good catch! I'll edit the reply and make note of that.
